aten/src/ATen/native/cuda/SummaryOps.cu

```diff
@@ -0,0 +1,284 @@
+         block,                                                            \
+         (MEMORY_TYPE == CUDAHistogramMemoryType::SHARED) ? sharedMem : 0, \
+         at::globalContext().getCurrentCUDAStream()>>>(                    \
+          aInfo, pInfo, bInfo, binsize, totalElements, WEIGHTS_OP);        \
+  AT_ASSERTM(cudaGetLastError() == cudaSuccess, "kernelHistogram1D failed");
+
+#define HANDLE_SWITCH_CASE(mType, getOp)                        \
+  switch (mType) {                                              \
+    case CUDAHistogramMemoryType::SHARED:                       \
+      HANDLE_CASE(CUDAHistogramMemoryType::SHARED, getOp);      \
+      break;                                                    \
+    case CUDAHistogramMemoryType::MULTI_BLOCK:                  \
+      HANDLE_CASE(CUDAHistogramMemoryType::MULTI_BLOCK, getOp); \
+      break;                                                    \
+    default:                                                    \
+      HANDLE_CASE(CUDAHistogramMemoryType::GLOBAL, getOp);      \
+  }
+
+/*
+  Calculate the frequency of the input values.
+
+  `a` contains the final output or the histogram.
+  Input `b` is assumed to be 1-D non-negative int array.
+  `c` optionally contains the weight vector.
+  See `help torch.bincount` for details on the math.
+
+  3 implementations based of input size and memory usage:
+    case: #bins < blockDim.x
+        SHARED: Each block atomically adds to it's own **shared** hist copy,
+        then atomically updates the global tensor.
+    case: blockDim.x <= #bins < MIN_NUMBER_BINS_FOR_GLOBAL_MEM
+        MULTI_BLOCK: Each block atomically adds to it's own **global** hist
+        copy, then atomically updates the global tensor.
+    case: MIN_NUMBER_BINS_FOR_GLOBAL_MEM <= #bins
+        GLOBAL: all threads atomically update to a single **global** hist copy.
+ */
+template <typename output_t, typename input_t, bool HasWeights>
+bool CUDA_tensor_histogram(
+    at::Tensor a, /* output */
+    at::Tensor b, /* input */
+    at::Tensor c, /* weights(optional) */
+    int64_t nbins,
+    int binsize,
+    TensorArgType aType = TensorArgType::ReadWrite,
+    TensorArgType bType = TensorArgType::ReadOnly,
+    TensorArgType cType = TensorArgType::ReadOnly) {
+  checkBackend("CUDA_tensor_histogram", {a, b}, Backend::CUDA);
+  if (HasWeights) {
+    checkBackend("CUDA_tensor_histogram", {c}, Backend::CUDA);
+  }
+  auto totalElements = b.size(0);
+
+  const dim3 block = getApplyBlock();
+  dim3 grid;
+  int64_t curDevice = current_device();
+  if (curDevice == -1)
+    return false;
+  if (!getApplyGrid(totalElements, grid, curDevice)) {
+    return false;
+  }
+#if CUDA_VERSION < 9000
+  grid.x = std::min(
+      (unsigned int)at::globalContext()
+              .getCurrentDeviceProperties()
+              ->multiProcessorCount *
+          AT_APPLY_BLOCKS_PER_SM,
+      grid.x);
+#endif
```

Comment:
What is going on here? (Why are we checking CUDA_VERSION?)

Reply:
I had picked this up from CUDAApplyUtils.cuh (which has now been removed) - I'll remove it.